desktop/cmp/treemap/TreeMap.js

```diff
@@ -0,0 +1,297 @@
+        if (this._clickCount >= 2) this._debouncedClickHandler.flush();
+    };
+
+    clickHandler(record, e) {
+        const {onClick, onDoubleClick} = this.model;
+        if (onClick && this._clickCount === 1) {
+            onClick(record, e);
+        } else if (onDoubleClick && this._clickCount === 2) {
+            onDoubleClick(record, e);
+        }
+        this._clickCount = 0;
+    }
+
+    //----------------------
+    // Selection handling
+    //----------------------
+    syncSelection() {
+        if (!this._chart) return;
+
+        const {selectedIds, maxDepth, gridModel} = this.model;
+        let toSelect = [...selectedIds];
+
+        // Fallback to parent node if selection exceeds max depth
+        if (maxDepth && gridModel && gridModel.treeMode) {
+            toSelect = selectedIds.map(id => {
+                const delim = Cube.RECORD_ID_DELIMITER,
+                    parts = id.split(delim),
+                    idDepth = parts.filter(p => p !== 'root').length,
+                    toRemove = idDepth - maxDepth;
+
+                return toRemove < 1 ? id : parts.slice(0, -toRemove).join(delim);
+            });
+        }
+
+        // Update selection in chart
+        this._chart.series[0].data.forEach(node => {
+            node.select(toSelect.includes(node.id), true);
+        });
+    }
+
+    //----------------------
+    // Labels
+    //----------------------
+    async updateLabelVisibilityAsync() {
+        if (!this._chart) return;
+
+        // Show / hide labels by comparing label size to node size
+        await forEachAsync(this._chart.series[0].data, node => {
+            if (node.dataLabel) {
+                const buffer = 10,
+                    tooSmallWidth = (node.dataLabel.width + buffer) > node.graphic.element.width.baseVal.value,
+                    tooSmallHeight = (node.dataLabel.height + buffer) > node.graphic.element.height.baseVal.value,
+                    style = tooSmallWidth || tooSmallHeight ? {visibility: 'hidden'} : {visibility: 'visible'};
+
+                node.update({dataLabels: {style}}, false, false);
+            }
+        });
+
+        this._chart.redraw();
+    }
+
+    //----------------------
+    // Tooltip
+    //----------------------
+    defaultTooltip = (record) => {
+        const {labelField, valueField, heatField, valueFieldLabel, heatFieldLabel} = this.model,
+            value = record[valueField],
+            name = record[labelField],
+            heat = record[heatField];
+
+        return `
+            <div class='xh-treemap-tooltip'>
+                <div class='xh-treemap-tooltip__label'>${name}</div>
+                <div class='xh-treemap-tooltip__row'>
+                    <div>${valueFieldLabel || valueField}:</div>
+                    <div>${fmtNumber(value)}</div>
+                </div>
+                <div class='xh-treemap-tooltip__row' ${valueField == heatField ? 'style="display:none"' : ''}>
```

Comment:
Would avoid writing this out at all (including the call to fmtNumber) in the no-op case where we won't be displaying a second row.